### PR TITLE
Ignore IsNotFound error

### PR DESCRIPTION
**Pull request by FengyunPan**
IsNotFound error is fine since that means the object is
deleted already, so we should check err and ignore err
before returning.

### CODE REVIEW

**Comment by quinton-hoole** [comment]
@k8s-bot ok to test

**Comment by quinton-hoole** [comment]
This looks good to me.  @marun would you like to do a quick second review, as you've had your head deep in this stuff lately.

**Comment by quinton-hoole** [comment]
Somehow my LGTM also approved this, which is not what I intended.  Waiting for a second review before approving.

File: federation/pkg/federation-controller/sync/controller.go [hunk 1]
```diff
@@ -176,9 +176,15 @@ func newFederationSyncController(client federationclientset.Interface, adapter f
 			return err
 		},
 		func(client kubeclientset.Interface, obj pkgruntime.Object) error {
+			secret := obj.(*apiv1.Secret)
```
Comment (marun) [inline federation/pkg/federation-controller/sync/controller.go]: This controller is generic - it's not intended to be just for secrets.  And a secret isn't needed here - that's what namespacedName is for.

File: federation/pkg/federation-controller/sync/controller.go [hunk 2]
```diff
@@ -176,9 +176,15 @@ func newFederationSyncController(client federationclientset.Interface, adapter f
...
 			namespacedName := adapter.NamespacedName(obj)
 			orphanDependents := false
 			err := adapter.ClusterDelete(client, namespacedName, &metav1.DeleteOptions{OrphanDependents: &orphanDependents})
+			// IsNotFound error is fine since that means the object is deleted already.
+			if errors.IsNotFound(err) {
+				glog.V(4).Infof("Secret %s/%s no longer exists", secret.Namespace, secret.Name)
```
Comment (marun) [inline federation/pkg/federation-controller/sync/controller.go]:
```golang
glog.V(4).Infof("%s %q no longer exists",  strings.ToTitle(adapter.Kind()), namespacedName)
```
The name of the type being federated can be retrieved via ``adapter.Kind()`` (returns lowercase), and namespacedName's ``String`` method results in ``[namespace]/[name]``.

**Review by marun — COMMENTED** [review]
Consider updating the federated updater to check the error for ``IsNotFound`` instead of making the change per-type.  The logging isn't really necessary since a NotFound error is safe to ignore in this case.

File: federation/pkg/federation-controller/sync/controller.go (hunk 1 above)
Comment (FengyunPan) [inline federation/pkg/federation-controller/sync/controller.go]: Done

File: federation/pkg/federation-controller/sync/controller.go (hunk 2 above)
Comment (FengyunPan) [inline federation/pkg/federation-controller/sync/controller.go]: Nice! Done.

**Comment by marun** [comment]
As per my review comment, why is it preferable to implement controller-specific logic?  Why not simply ignore ``NotFound`` in [FederatedUpdater.UpdateWithOnError](https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/federation-controller/util/federated_updater.go#L96):

```golang
switch op.Type {
...
case OperationTypeDelete:
  err = fu.deleteFunction(clientset, op.Obj)
  if errors.IsNotFound(err) {
    err = nil
  }
}
```

**Comment by marun** [comment]
@k8s-bot non-cri e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

File: federation/pkg/federation-controller/sync/controller.go [hunk 3]
```diff
@@ -115,7 +115,7 @@ func newFederationSyncController(client federationclientset.Interface, adapter f
 		adapter:               adapter,
 	}
 
-	// Build delivereres for triggering reconciliations.
+	// Build deliverers for triggering reconciliations.
```
Comment (marun) [inline federation/pkg/federation-controller/sync/controller.go]: Please separate these unrelated and non-functional changes into a separate commit.

**Comment by marun** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by madhusudancs** [comment]
@k8s-bot kops aws e2e test this

**Comment by madhusudancs** [comment]
@FengyunPan could you please ping me once all the tests pass here - http://prow.k8s.io/?type=presubmit&repo=kubernetes%2Fkubernetes&pull=44483

**Comment by FengyunPan** [comment]
@madhusudancs yes.

**Comment by FengyunPan** [comment]
@madhusudancs Hi, the pull-kubernetes-e2e-kubeadm-gce is FAILURE, but I think that is not the result of this patch.

**Comment by madhusudancs** [comment]
@FengyunPan kubeadm test is not required, so we can ignore that.

We need federation tests to pass for federation PRs and it failed last time. I am going to run it again. You can monitor the progress at the same URL.

@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by FengyunPan** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot kops aws e2e test this

**Comment by marun** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

Results: http://prow.k8s.io/?type=presubmit&pull=44483&job=pull-kubernetes-federation-e2e-gce

**Comment by madhusudancs** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by madhusudancs** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by madhusudancs** [comment]
Federation presubmit passed - https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/44483/pull-kubernetes-federation-e2e-gce/2698/

**Pull request merged**